### PR TITLE
[Serve] Add gRPC context related docs

### LINKED ISSUE #41671: [Serve] Add docs related to using gRPC context in Serve
### Description

In Ray 2.9 we added new feature to use grpc context in deployments. Let's also add related doc and examples.

### Link

_No response_

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

Adding example code and descriptions for using gRPC context in Serve.

## Related issue number

Closes https://github.com/ray-project/ray/issues/41671

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/serve/advanced-guides/grpc-guide.md [hunk 1]
```diff
@@ -290,3 +291,32 @@ Serve uses the following gRPC error codes:
   this node and you should no longer route to this node.
 - `DEADLINE_EXCEEDED`: The request took longer than the timeout setting and got cancelled.
 - `INTERNAL`: Other unhandled errors during the request.
+
+(serve-grpc-proxy-grpc-context)=
+## Use gRPC context
+Serve provides a gRPC context object to the deployment replica to get information
```
Comment (edoakes) [inline doc/source/serve/advanced-guides/grpc-guide.md:297]: Suggest to link to the context object in the gRPC API reference

File: doc/source/serve/advanced-guides/grpc-guide.md [hunk 2]
```diff
@@ -290,3 +291,32 @@ Serve uses the following gRPC error codes:
...
+details, and trailing metadata.
+
+```{literalinclude} ../doc_code/grpc_proxy/grpc_guide.py
+:start-after: __begin_grpc_context_define_app__
+:end-before: __end_grpc_context_define_app__
+:language: python
+```
+
+We can define the client code like the following to get the custom code, details, and
+trailing metadata.
```
Comment (edoakes) [inline doc/source/serve/advanced-guides/grpc-guide.md:311]: This is redundant with the above. Maybe just remove it, or else "refactor" and remove the last sentence from the above paragraph and only have the information here.

File: doc/source/serve/advanced-guides/grpc-guide.md [hunk 3]
```diff
@@ -290,3 +291,32 @@ Serve uses the following gRPC error codes:
...
+```{literalinclude} ../doc_code/grpc_proxy/grpc_guide.py
+:start-after: __begin_grpc_context_client__
+:end-before: __end_grpc_context_client__
+:language: python
+```
+
+:::{note}
+If the deployment errors out for any reason, Serve will continue to return
+`INTERNAL` code with the stacktrace in the details, regardless of what code and details
+are set in the `RayServegRPCContext` object.
```
Comment (edoakes) [inline doc/source/serve/advanced-guides/grpc-guide.md:321]:
```suggestion
If the handler raises an unhandled exception, Serve will return an
`INTERNAL` error code with the stacktrace in the details, regardless of what code and details
are set in the `RayServegRPCContext` object.
```

File: doc/source/serve/doc_code/grpc_proxy/grpc_guide.py [hunk 4]
```diff
@@ -309,3 +309,85 @@ def __call__(self, image: Image):
...
+
+    def __call__(
+        self,
+        user_message: UserDefinedMessage,
+        grpc_context: RayServegRPCContext,  # to use grpc context, add this kwarg
+    ) -> UserDefinedResponse:
+        greeting = f"Hello {user_message.name} from {user_message.origin}"
+        num, code, message = self.num_lookup(user_message.name)
+
+        # Use custom code, details, and trailing metadata
```
Comment (edoakes) [inline doc/source/serve/doc_code/grpc_proxy/grpc_guide.py:347]:
```suggestion
        # Set custom code, details, and trailing metadata.
```

**Review by edoakes — COMMENTED** [review]
Suggestions to make the text more concise

**Comment by GeneDer** [comment]
@edoakes I address all the comments, can you PTAL again when you have a sec

**Review by edoakes — APPROVED** [review]

**Pull request merged**